### PR TITLE
Added selecting virtualservice

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Please provide a description for what this PR is for.

Only handle virtualServices related to this route, instead of all

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
fix #33760

File: pilot/pkg/networking/core/v1alpha3/httproute_test.go [hunk 1]
```diff
@@ -957,9 +996,6 @@ func TestSidecarOutboundHTTPRouteConfig(t *testing.T) {
 				"test-private.com:80": {
 					"test-private.com": true, "test-private.com:80": true, "9.9.9.9": true, "9.9.9.9:80": true,
 				},
-				"test-private-3.com:80": {
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/httproute_test.go:960]: can you add some explanation? This seems like a major breaking change

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/httproute_test.go:960]: Oh nevermind, I see the linked issue now

**Review by howardjohn — COMMENTED** [review]
I am not sure what problems this is solving. What is the problem with adding the route?

**Comment by hzxuzhonghu** [comment]
There are two caveats:
1. for each route, unknown vses are used to build http route
2.  only for route 80, the virtualhost includes the unknown hosts. (It is not intended, but there are some tests depend on this behavior)

**Comment by hzxuzhonghu** [comment]
IMO, we can leave the hacky 80 only for listenerPort=80 to not break compatibility if someone depends on this behavior, and reduce the cpu cost

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 2]
```diff
@@ -199,15 +199,11 @@ func buildSidecarVirtualHostsForVirtualService(
...
-	// going to send a separate RDS request
-	// Note that we need to build non-default HTTP routes only for the virtual services.
-	// The services in the serviceRegistry will always have a default route (/)
 	if len(serviceByPort) == 0 {
-		// This is a gross HACK. Fix me. Its a much bigger surgery though, due to the way
-		// the current code is written.
-		serviceByPort[80] = nil
+		// no services or no service port for the listener port
+		return nil
 	}
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:205]: This is the legacy hack code

**Comment by ramaraochavali** [comment]
This change makes sense to me

**Comment by howardjohn** [comment]
>  (It is not intended, but there are some tests depend on this behavior)

Are these unit tests? Or the security integration tests that are failing?

If its integration tests I would be worried this would be impacting users. For unit tests maybe not, since unit tests sometimes test impossible configurations

**Comment by hzxuzhonghu** [comment]
The the security integration tests depends on this.


Before this patch, even no SE defined for `www.google.com`, the eventual output is: 
1.  one VirtualHost generated for `google.com` on route 80. 
2.  there are also VirtualHost built for each route as an intermediate procedure, but not added at all. 

If you have concern of breaking some compatibility, we can leave 1, and probihit  2. 

In this way it is safer, there is no user-sensitive behavior.

```
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: test
spec:
  hosts:
  - www.google.com
  http:
  - name: "test"
    route:
    - destination:
        host: www.google.com
```

**Comment by hzxuzhonghu** [comment]
@ramaraochavali @howardjohn Reverted the hack behavior for 80

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 3]
```diff
@@ -230,6 +270,11 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundVirtualHosts(node *mod
 		}
 	}
 
+	// This is hack to keep consistent with previous behavior.
+	if listenerPort != 80 {
+		// only select virtualServices that matches a service
+		virtualServices = selectVirtualServices(virtualServices, servicesByName)
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/httproute.go:276]: I do not think we have any test coverage of this change. If you comment it out all tests still pass. 

This also makes it hard to review the rest as I don't understand the full scope of the behavior changed by the PR, so I am not sure if this is safe or not. We would want to make sure these routes we are adding are truly never used before removing them, as we have had a lot of regressions in this config in the past

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:276]: Yup, it passes without this selection. I explained above https://github.com/istio/istio/pull/33764#issuecomment-873505981
>there are also VirtualHost built for each route as an intermediate procedure, but not added at all.

I am thinking about adding a test for this, but does not come up with a good way

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:276]: addes test

**Comment by hzxuzhonghu** [comment]
@howardjohn @ramaraochavali ptal, will depend on this to do rds cache, otherwise, there will be too many vses

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 4]
```diff
@@ -181,6 +181,46 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundHTTPRouteConfig(node *
 	return out
 }
 
+// TODO: merge with IstioEgressListenerWrapper.selectVirtualServices
+func selectVirtualServices(virtualServices []config.Config, servicesByName map[host.Name]*model.Service) []config.Config {
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/httproute.go:185]: nit: Can you add a comment on what this does?

**Review by ramaraochavali — COMMENTED** [review]
LGTM.Will keep it open for @howardjohn review. I will approve on Monday.

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:185]: yes, make sense

**Comment by hzxuzhonghu** [comment]
ping @howardjohn

**Review by ramaraochavali — APPROVED** [review]
@howardjohn I am approving this as it LGTM. If you have further comments, @hzxuzhonghu can do a followup.

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:277]: Removing this still passes all tests I think

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:277]: Oh but we have tests for the function itself, i see

**Pull request merged**